File: 프로그래머스/2/92342. 양궁대회/양궁대회.java
import java.util.*;

class Solution {
    static int[] inputInfo;
    static int[] answer;
    static int total=0;

    public static int[] solution(int n, int[] info) {
        inputInfo = info;
        answer = new int[]{-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1};
        dfs(0, n, new int[11]);
        if (answer[0] == -1) {
            return new int[]{-1};
        }
        return answer;
    }

    public static void dfs(int cur, int n, int[] points) {
        if (cur == n) {
            int R = 0;
            int A = 0;
            for (int i = 0; i < points.length; i++) {
                if (points[i] > inputInfo[i]) {
                    R += (10 - i);
                } else {
                    if (points[i] == 0 && inputInfo[i] == 0) {
                        continue;
                    } else {
                        A += (10 - i);
                    }
                }
            }
            if (R > A) {
                if (answer[0] == -1) {
                    for(int i=0; i<points.length; i++){
                        answer[i] = points[i];
                        total = R-A;
                    }
                } else {
                    if(total>R-A){
                        return;
                    }
                    if(total<R-A){
                        for(int i=0; i<points.length; i++){
                            answer[i] = points[i];
                            total = R-A;
                        }
                        return;
                    }
                    for (int i = 10; i >= 0; i--) {
                        if (answer[i] > points[i]) {
                            break;
                        }
                        if (answer[i] < points[i]) {
                            for(int j=0; j<points.length; j++){
                                answer[j] = points[j];
                            }
                            break;
                        }
                    }
                }
            }
            return;
        }
        for (int i = 0; i < 11; i++) {
            if(points[i] <  inputInfo[i] && cur+inputInfo[i]+1 <=n){
                points[i] += inputInfo[i]+1;
                dfs(cur+inputInfo[i]+1, n, points);
                points[i] -= inputInfo[i]+1;
            }
            else{
                if(points[i] == inputInfo[i]){
                    points[i]++;
                    dfs(cur + 1, n, points);
                    points[i]--;
                }
                else{
                    points[n]+= n-cur;
                    dfs(n, n, points);
                    points[n]-= n-cur;
                }
            }
        }
    }
}

File: 프로그래머스/2/92342. 양궁대회/양궁대회.py
# k 점 더많이 맞혔을 경우 더많이 맟힌 선수가 k점 가져감
# a==b 어피치가 k점 가져감 -> 나는 라이언
# a==b==0 아무도 점수 X
# 주의 사항 k점만 가져간다 여러번 X
# 최종점수가 같으면 어피치 우승자
#info 0이라면 10점 가장 큰 점수차 로 결과값 없다면  [-1] 리턴
# 어차피 이기려면 부조건 어퍼치보다 더 많이 먹어야함
from collections import deque
def solution(n, info):
    queue = deque()
    firstSum = 0
    for i in range(len(info)):
        if info[i] != 0:
            firstSum += (10-i)
    queue.append((0,0,n,[]))
    answer = [-1]
    countAns = []
    ansSum = firstSum
    while queue:
        index,countSum,left,entry = queue.popleft()
        if len(entry) == 11 and countSum > firstSum and countSum >= ansSum:
            if ansSum == countSum:
                countAns.append(entry[0:10] + [entry[10] + left])
            else:
                ansSum = countSum
                countAns = []
                countAns.append(entry[0:10] + [entry[10] + left])
        if index > len(info)-1:
            continue
        queue.append((index+1,countSum,left,entry+[0]))
        if info[index] >= left:
            continue
        if info[index] != 0:
            queue.append((index+1,countSum + 2*(10 - index),left-info[index]-1,entry+[info[index]+1]))
        else:
            queue.append((index+1,countSum + 10 - index,left-info[index]-1,entry+[info[index]+1]))
    countAns.sort(reverse= True,key=lambda x: x[::-1])
    if len(countAns) == 0:
            return [-1]     
    return countAns[0]